Src/Plugins/Security/SEC101_054.AlibabaAccessKeyValidator.cs

```diff
@@ -0,0 +1,119 @@
+        }
+
+        public static string IsValidDynamic(ref string fingerprint, ref string message)
+        {
+            return IsValidDynamic(Instance,
+                                   ref fingerprint,
+                                   ref message);
+        }
+
+        protected override string IsValidStaticHelper(ref string matchedPattern,
+                                                      ref Dictionary<string, string> groups,
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("account", out string account) ||
+                !groups.TryGetNonEmptyValue("password", out string password))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint()
+            {
+                Account = account,
+                Password = password,
+            }.ToString();
+
+            return nameof(ValidationState.AuthorizedWarning);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText,
+                                                       ref string message)
+        {
+            Fingerprint fingerprint = new Fingerprint(fingerprintText);
+            string account = fingerprint.Account;
+            string password = fingerprint.Password;
+
+            try
+            {
+                // Taken from https://www.alibabacloud.com/help/doc-detail/63638.htm
+                IClientProfile clientProfile = DefaultProfile.GetProfile("cn-shanghai", account, password);
+                DefaultAcsClient client = new DefaultAcsClient(clientProfile);
+
+                PubRequest request = new PubRequest();
+
+                // We don't actually care about any of the product stuff, so just leave dummy values in
+                request.ProductKey = "<productKey>";
+                request.TopicFullName = "/<productKey>/<deviceName>/get";
+                byte[] payload = Encoding.Default.GetBytes("Hello World.");
+                string payloadStr = Convert.ToBase64String(payload);
+                request.MessageContent = payloadStr;
+                request.Qos = 0;
+                PubResponse response = client.GetAcsResponse(request);
+
+                // If all goes well, we'll receive a "product format invalid" message in the response
+                // which means authentication succeeded.  Therefore the id and secret are valid.
+                return ReturnAuthorizedAccess(ref message, asset: account);
+            }
+            catch (ClientException ce)
+            {
+                switch (ce.ErrorCode)
+                {
+                    case KeyNotFound:
+                        // Not even the client id we found is valid.  Return no match.
```

Comment:
>   [](start = 68, length = 2)

nit: remove the double space and in line 110

Reply:
Done

---
In reply to: [591817637](https://github.com/microsoft/sarif-pattern-matcher/pull/285#discussion_r591817637) [](ancestors = 591817637)